tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -2521,5 +2822,1033 @@ bool StitchCPUAnalysis::doCodeGeneration(OpBuilder& b, lmhlo::FusionOp fusion) {
+  auto results = fusion_pattern.getResults();
+  for (auto res : results) {
+    bool irregular = false;
+    Operation* op = fusion_pattern.findLastWriter(res);
+    if (subroots_set.contains(op)) {
+      continue;
+    }
+    // Find equal dims between res and a sub-root.
+    SmallVector<std::pair<SmallVector<int64_t>, SmallVector<int64_t>>> equal;
+    shapeAnalysis.extractContDimEqualInfo(dominant, res, equal);
+    int64_t dominant_non_tiled_dim_matched_n = 0;
+    DenseSet<int64_t> res_non_tiled_dims;
+    for (auto eq_dim : equal) {
+      auto lhs = eq_dim.first;
+      bool non_tiled = (dominant_tile.tileSizes.count(lhs[0]) == 0);
+      for (int64_t i = 1; i < lhs.size(); i++) {
+        if ((dominant_tile.tileSizes.count(lhs[i]) == 0) != non_tiled) {
+          // Both tiled and non-tiled dims are mapped to the same dim of res,
+          // thus cannot determine the tile plan for res;
+          irregular_xroots.insert(op);
+          irregular = true;
+          break;
+        }
+      }
+      if (irregular) {
+        break;
+      }
+      if (non_tiled) {
+        res_non_tiled_dims.insert(eq_dim.second.begin(), eq_dim.second.end());
+        dominant_non_tiled_dim_matched_n += lhs.size();
+      }
+    }
+    if (irregular) {
+      continue;
+    } else if (dominant_non_tiled_dim_matched_n != 1) {
+      // Check constraint: the element-number of non-tiled dims are the same
+      // between sub-roots and regular xroots. If not every non-tiled dim is
+      // matched between sub-root and current op, it is an irregular xroot. Note
+      // we know that sub-root is reduction and has only 1 non-tiled dim.
+      irregular_xroots.insert(op);
+      continue;
+    } else if (*std::max_element(res_non_tiled_dims.begin(),
+                                 res_non_tiled_dims.end()) !=
+               res_non_tiled_dims.size() - 1) {
+      // Check constraint: the tiled dims are all minor dims for all regular
+      // xroot ops. This means the non-tiled dims should all be the major dims.
+      irregular_xroots.insert(op);
+      continue;
+    }
+
+    auto& plan = tile_plan[res];
+    int64_t rank = res.getType().cast<MemRefType>().getRank();
+    for (int64_t i = 0; i < rank; i++) {
+      if (res_non_tiled_dims.count(i) == 0) {
+        plan.tileSizes[i] = ShapedType::kDynamicSize;
+      }
+    }
+    regular_xroots.insert(op);
+  }
+
+  // Check constraint: all the external-only-roots should be regular. This is
+  // because all external-only-roots are skeleton ops. It requires all skeleton
+  // ops have the same non-tiling element numbers to ease the building of loop
+  // loop skeleton of the fusion during code generation.
+  for (auto extern_only_res : fusion_pattern.getExternalOnlyResults()) {
+    Operation* op = fusion_pattern.findLastWriter(extern_only_res);
+    if (irregular_xroots.contains(op)) {
+      LLVM_DEBUG(llvm::dbgs() << "Invalid irregular xroot: " << *op);
+      return false;
+    }
```

Comment:
This constraint sounds weird to me. A result op with internal users can be supported as irregular roots, while an external only result cannot on the contrary?

Reply:
Irregular means the tiling information is unknown, which cannot be used as skeleton op. An external only root op has to be a skeleton op, which cannot be irregular.